controllers/propagator/template_utils.go

```diff
@@ -0,0 +1,647 @@
+package propagator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	templates "github.com/stolostron/go-template-utils/v6/pkg/templates"
+	k8sdepwatches "github.com/stolostron/kubernetes-dependency-watches/client"
+	authv1 "k8s.io/api/authentication/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	policiesv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
+	"open-cluster-management.io/governance-policy-propagator/controllers/common"
+)
+
+var (
+	// A special NamespacedName to indicate the default service account is used
+	defaultSANamespacedName types.NamespacedName = types.NamespacedName{Name: "<default>"}
+	// A label selector to select policies that have a label indicating the parent/root policy.
+	replicatedPlcLabelSelector labels.Selector
+)
+
+func init() {
+	replicatedReq, err := labels.NewRequirement(common.RootPolicyLabel, selection.Exists, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	replicatedPlcLabelSelector = labels.NewSelector().Add(*replicatedReq)
+}
+
+type TokenRefreshConfig struct {
+	// The token lifetime in seconds.
+	ExpirationSeconds int64
+	// The minimum refresh minutes before expiration. This must be <= MaxRefreshMins.
+	MinRefreshMins float64
+	// The maximum refresh minutes before expiration. This must be >= MinRefreshMins.
+	MaxRefreshMins float64
+	// If a token refresh encountered an unrecoverable error, then this is called.
+	OnFailedRefresh func(error)
+}
+
+type templateResolverWithCancel struct {
+	// cancel will stop the TemplateResolver and the automatic token refresh.
+	cancel context.CancelFunc
+	// canceled will be set if resolver failed to start or was stopped. The atomic.Bool type is used to avoid using
+	// a write lock.
+	canceled atomic.Bool
+	// referenceCount is a counter for how many replicated policies rely on this template resolver. Once it is 0,
+	// the template resolver will be cleaned up. A pointer is used since the reference count can be copied to a new
+	// templateResolverWithCancel instance if the TemplateResolver instance was unexpectedly canceled and was
+	// restarted.
+	referenceCount *atomic.Uint32
```

Comment:
I can't shake the feeling that us doing our own reference counting means we're missing some other way to do this

Reply:
The only other way I could think of that would work other than periodically checking to see if any root policy still references this service account was to have a map of service accounts to replicated policies. I thought the reference count was better to avoid locking around a shared map. It also saves a bit of memory.

Another way that came to mind was using a read and write lock for each template resolver and all replicated policies have a read lock and then a separate goroutine waits for a write lock, which would indicate there are no more readers left to then clean up the template resolver. The problem is that the read and write lock in the Go standard library doesn't work like that as it prioritizes giving the write lock when it can, which makes sense.